src/serac/numerics/functional/tuple_tensor_dual_functions.hpp

```diff
@@ -0,0 +1,673 @@
+{
+  double x, df_dx;
+  double fl = f(lower_bound, get_value(params)...);
+  double fh = f(upper_bound, get_value(params)...);
+
+  if (fl * fh > 0) {
+    SLIC_ERROR("solve_scalar_equation: root not bracketed by input bounds.");
+  }
+
+  unsigned int iterations = 0;
+  bool         converged  = false;
+
+  // handle corner cases where one of the brackets is the root
+  if (fl == 0) {
+    x         = lower_bound;
+    converged = true;
+  } else if (fh == 0) {
+    x         = upper_bound;
+    converged = true;
+  }
+
+  if (converged) {
+    df_dx = get_gradient(f(make_dual(x), get_value(params)...));
+
+  } else {
+    // orient search so that f(xl) < 0
+    double xl = lower_bound;
+    double xh = upper_bound;
+    if (fl > 0) {
+      xl = upper_bound;
+      xh = lower_bound;
+    }
+
+    // move initial guess if it is not between brackets
+    if (x0 < lower_bound || x0 > upper_bound) {
+      x0 = 0.5 * (lower_bound + upper_bound);
+    }
+    const unsigned int MAX_ITERATIONS = 25;
+    x                                 = x0;
+    double delta_x_old                = std::abs(upper_bound - lower_bound);
+    double delta_x                    = delta_x_old;
+    auto   R                          = f(make_dual(x), get_value(params)...);
+    auto   fval                       = get_value(R);
+    df_dx                             = get_gradient(R);
+
+    while (!converged) {
+      if (iterations == MAX_ITERATIONS) {
+        SLIC_WARNING("solve_scalar_equation failed to converge in allotted iterations.");
+        break;
+      }
+
+      // use bisection if Newton oversteps brackets or is not decreasing sufficiently
+      if ((x - xh) * df_dx - fval > 0 || (x - xl) * df_dx - fval < 0 ||
+          std::abs(2. * fval) > std::abs(delta_x_old * df_dx)) {
+        delta_x_old = delta_x;
+        delta_x     = 0.5 * (xh - xl);
+        x           = xl + delta_x;
+        converged   = (x == xl);
+      } else {  // use Newton step
+        delta_x_old = delta_x;
+        delta_x     = fval / df_dx;
+        auto temp   = x;
+        x -= delta_x;
+        converged = (x == temp);
+      }
+
+      // std::cout << "iter " << iterations << " x = " << x << std::endl;
+
+      // convergence check
+      converged = converged || (std::abs(delta_x) < tolerance);
+
+      // function and jacobian evaluation
+      R     = f(make_dual(x), get_value(params)...);
+      fval  = get_value(R);
+      df_dx = get_gradient(R);
+
+      // maintain bracket on root
+      if (fval < 0) {
+        xl = x;
+      } else {
+        xh = x;
+      }
+
+      ++iterations;
+    }
+  }
+
+  // Accumulate derivatives so that the user can get derivatives
+  // with respect to parameters, subject to constraing that f(x, p) = 0 for all p
+  // Conceptually, we're doing the following:
+  // [fval, df_dp] = f(get_value(x), p)
+  // df = 0
+  // for p in params:
+  //   df += inner(df_dp, dp)
+  // dx = -df / df_dx
+  constexpr bool contains_duals =
```

Comment:
Is this useful enough to be it's own function? Essentially is anything in the parameter pack a dual?

Reply:
> Is this useful enough to be it's own function?

I'd say not yet. Brandon and I noticed that these type traits like `is_dual` are almost never used in serac, so there's not a huge demand for them right now.